spec/app/domain/authentication/authn-azure/validate_application_identity_spec.rb

```diff
@@ -0,0 +1,428 @@
+
+  let(:resource_class) { double("Resource") }
+  let(:host) { double ("some-host") }
+  let(:non_existent_host_id) { double ("some-non-existent-host-id") }
+  let(:xms_mirid_token_field_user_assigned_identity) { "/subscriptions/some-subscription-id-value/resourcegroups/some-resource-group-value/providers/Microsoft.ManagedIdentity/userAssignedIdentities/some-user-assigned-identity-value" }
+  let(:xms_mirid_token_field_system_assigned_identity) { "/subscriptions/some-subscription-id-value/resourcegroups/some-resource-group-value/providers/Microsoft.Compute/virtualMachines/some-system-assigned-identity-value" }
+
+  let(:xms_mirid_token_missing_subscription_id_field) { "/resourcegroups/some-resource-group-value/providers/Microsoft.ManagedIdentity/userAssignedIdentities/some-system-assigned-identity-value" }
+  let(:xms_mirid_token_missing_resource_groups_field) { "/subscriptions/some-subscription-id-value/providers/Microsoft.ManagedIdentity/some-user-assigned-identity-value" }
+  let(:xms_mirid_token_missing_providers_field) { "/subscriptions/some-subscription-id-value/resourcegroups/some-resource-group-value/" }
+  let(:xms_mirid_token_missing_initial_slash) { "subscriptions/some-subscription-id-value/resourcegroups/some-resource-group-value/providers/Microsoft.ManagedIdentity/userAssignedIdentities/some-system-assigned-identity-value" }
+
+  let(:invalid_xms_mirid_token_providers_field) { "/subscriptions/some-subscription-id-value/resourcegroups/some-resource-group-value/providers/Microsoft.ManagedIdentity/some-user-assigned-identity-value" }
+
+  let(:oid_token_field) { "test-oid" }
+
+  let(:subscription_id_annotation) { double("SubscriptionIdAnnotation") }
+  let(:resource_group_annotation) { double("ResourceGroupAnnotation") }
+  let(:user_assigned_identity_annotation) { double("UserAssignedIdentityAnnotation") }
+  let(:system_assigned_identity_annotation) { double("SystemAssignedIdentityAnnotation") }
+
+  let(:mismatched_subscription_id_annotation) { double("MismatchedSubscriptionIdAnnotation") }
+  let(:mismatched_resource_group_annotation) { double("MismatchedResourceGroupAnnotation") }
+  let(:mismatched_system_assigned_identity_annotation) { double("MismatchedSystemAssignedAnnotation") }
+  let(:mismatched_user_assigned_identity_annotation) { double("MismatchedUserAssignedAnnotation") }
+
+  let(:validate_azure_annotations) { double("ValidateAzureAnnotations") }
+
+  def mock_annotation_builder(host_annotation_type, host_annotation_key, host_annotation_value)
+    allow(host_annotation_type).to receive(:values)
+                                     .and_return(host_annotation_type)
+    allow(host_annotation_type).to receive(:[])
+                                     .with(:name)
+                                     .and_return(host_annotation_key)
+    allow(host_annotation_type).to receive(:[])
+                                     .with(:value)
+                                     .and_return(host_annotation_value)
+  end
```

Comment:
This method repeats itself. Looks like you can use a spec helper. Take a look at `oidc_spec_helper.rb` to see how we did it for OIDC. then move there all the shared code.

Reply:
good idea. added